From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mio <mio@mio19.uk>
Date: Tue, 20 Oct 2026 19:26:43 +1100
Subject: [PATCH] adapt the error UI to round smartwatch displays (boot/error)

Make the stage-1 error screen responsive to the configured display geometry so text and layout remain readable on round and chin-shaped smartwatch panels. Center the message and action panes inside the round display safe area and center the message text instead of laying them out from x=0.
---
 boot/error/lib/shaded_text.rb | 25 ++++++++++----
 boot/error/main.rb            | 61 ++++++++++++++++++++++++++---------
 2 files changed, 63 insertions(+), 23 deletions(-)

diff --git a/boot/error/lib/shaded_text.rb b/boot/error/lib/shaded_text.rb
index ee7ecd69..bbb43174 100644
--- a/boot/error/lib/shaded_text.rb
+++ b/boot/error/lib/shaded_text.rb
@@ -2,16 +2,27 @@
 class ShadedText
   OFFSET = 2
 
-  def initialize(*args)
+  def initialize(*args, size: nil, line_space: nil)
     @secondary = LVGL::LVLabel.new(*args)
-    style = @secondary.get_style(LVGL::LABEL_STYLE::MAIN).dup
-    @secondary.set_style(LVGL::LABEL_STYLE::MAIN, style)
-    style.text_color = 0xFF000000
+    sec_style = @secondary.get_style(LVGL::LABEL_STYLE::MAIN).dup
+    @secondary.set_style(LVGL::LABEL_STYLE::MAIN, sec_style)
+    sec_style.text_color = 0xFF000000
 
     @main = LVGL::LVLabel.new(*args)
-    style = @main.get_style(LVGL::LABEL_STYLE::MAIN).dup
-    @main.set_style(LVGL::LABEL_STYLE::MAIN, style)
-    style.text_color = 0xFFFFFFFF
+    main_style = @main.get_style(LVGL::LABEL_STYLE::MAIN).dup
+    @main.set_style(LVGL::LABEL_STYLE::MAIN, main_style)
+    main_style.text_color = 0xFFFFFFFF
+
+    if size
+      font = LVGUI::Fonts.primary(size)
+      sec_style.text_font = font
+      main_style.text_font = font
+    end
+
+    if line_space
+      sec_style.text_line_space = line_space
+      main_style.text_line_space = line_space
+    end
 
     # Ensures the position is reset at least once, to sync-up the shade.
     set_pos(0, 0)
diff --git a/boot/error/main.rb b/boot/error/main.rb
index e033ae02..487a45d4 100644
--- a/boot/error/main.rb
+++ b/boot/error/main.rb
@@ -43,9 +43,12 @@ class UI
 
   def add_title_bar()
     @title_bar = LVGL::LVContainer.new(@screen)
-    @title_bar.set_width(@screen.get_width())
-    @title_bar.set_height(16*unit)
-    @title_bar.set_pos(0, 0)
+    top_offset = round_screen? ? 5 * unit : 0
+    title_width = get_pane_width()
+    title_x = (@screen.get_width() - title_width) / 2
+    @title_bar.set_width(title_width)
+    @title_bar.set_height(round_screen? ? 10 * unit : 16 * unit)
+    @title_bar.set_pos(title_x, top_offset)
 
     @title_bar.get_style(LVGL::CONT_STYLE::MAIN).dup.tap do |style|
       @title_bar.set_style(LVGL::CONT_STYLE::MAIN, style)
@@ -56,7 +59,7 @@ class UI
       style.body_opa = (255 * 0.30).to_i
     end
 
-    add_sad_phone
+    add_sad_phone unless round_screen?
     add_title
   end
 
@@ -85,7 +88,7 @@ class UI
   def add_messages_pane()
     @messages_pane = LVGL::LVContainer.new(@screen)
     @messages_pane.set_width(get_pane_width())
-    @messages_pane.set_pos(0, @title_bar.get_height())
+    @messages_pane.set_pos(horizontal? ? 0 : pane_x, @title_bar.get_height())
 
     # When horizontal, full height
     if horizontal?
@@ -114,18 +117,26 @@ class UI
   end
 
   def add_title()
-    @title = new_text($code, parent: @title_bar)
-    @title.set_align(LVGL::LABEL_ALIGN::LEFT)
-    @title.set_x(@sad_phone.get_x()*2 + @sad_phone.get_width())
-    @title.set_y(@title_bar.get_height / 2 - @title.get_height / 2)
-    @title.set_width(@title_bar.get_width() - @title.get_x())
+    if round_screen?
+      @title = new_text($code, parent: @title_bar, size: 13)
+      @title.set_align(LVGL::LABEL_ALIGN::CENTER)
+      @title.set_x(@title_bar.get_width / 2 - @title.get_width / 2)
+      @title.set_y(@title_bar.get_height / 2 - @title.get_height / 2)
+      @title.set_width(@title_bar.get_width() - 2 * padding)
+    else
+      @title = new_text($code, parent: @title_bar)
+      @title.set_align(LVGL::LABEL_ALIGN::LEFT)
+      @title.set_x(@sad_phone.get_x()*2 + @sad_phone.get_width())
+      @title.set_y(@title_bar.get_height / 2 - @title.get_height / 2)
+      @title.set_width(@title_bar.get_width() - @title.get_x())
+    end
   end
 
   # Messages pane elements
 
   def add_message_title()
     @message_title = new_text($title, parent: @messages_pane)
-    @message_title.set_align(LVGL::LABEL_ALIGN::LEFT)
+    @message_title.set_align(text_align())
     @message_title.set_pos(
       padding,
       padding,
@@ -134,7 +145,7 @@ class UI
 
   def add_message()
     @message = new_text($message, parent: @messages_pane)
-    @message.set_align(LVGL::LABEL_ALIGN::LEFT)
+    @message.set_align(text_align())
     @message.set_pos(
       padding,
       @message_title.get_height() + @message_title.get_y() + padding
@@ -209,17 +220,17 @@ class UI
       )
 
       # Always push it as far down as possible!
-      @actions_pane.set_pos(0, @screen.get_height() - @actions_pane.get_height())
+      @actions_pane.set_pos(pane_x, @screen.get_height() - @actions_pane.get_height())
     end
   end
 
   # Misc. helpers
 
   # Creates a label with some useful defaults.
-  def new_text(text, parent: nil)
+  def new_text(text, parent: nil, size: nil, line_space: nil)
     parent ||= @screen
 
-    el = ShadedText.new(parent)
+    el = ShadedText.new(parent, size: size, line_space: line_space)
     el.set_long_mode(LVGL::LABEL_LONG::BREAK)
     el.set_align(LVGL::LABEL_ALIGN::CENTER)
     el.set_width((parent.get_width - 2*padding).to_i)
@@ -244,11 +255,21 @@ class UI
   def get_pane_width()
     if horizontal?
       @screen.get_width() / 2
+    elsif round_screen?
+      (@screen.get_width() * 0.74).to_i
     else
       @screen.get_width()
     end
   end
 
+  def pane_x()
+    (@screen.get_width() - get_pane_width()) / 2
+  end
+
+  def text_align()
+    round_screen? ? LVGL::LABEL_ALIGN::CENTER : LVGL::LABEL_ALIGN::LEFT
+  end
+
   def unit()
     (if horizontal?
       @screen.get_height()
@@ -257,8 +278,16 @@ class UI
     end) / 128
   end
 
+  def round_screen?()
+    @screen.get_width == @screen.get_height && @screen.get_width <= 500
+  end
+
   def padding()
-    2 * unit
+    if round_screen?
+      3 * unit
+    else
+      2 * unit
+    end
   end
 
   def horizontal?()
-- 
2.56.0

